elastica/modules/contact.py

```diff
@@ -0,0 +1,171 @@
+__doc__ = """
+Contact
+-------
+
+Provides the contact interface to apply contact forces between objects
+(rods, rigid bodies, surfaces).
+"""
+
+
+class Contact:
+    """
+    The Contact class is a module for applying contact between rod-like objects . To apply contact between rod-like objects,
+    the simulator class must be derived from the Contact class.
+
+        Attributes
+        ----------
+        _contacts: list
+            List of contact classes defined for rod-like objects.
+    """
+
+    def __init__(self):
+        self._contacts = []
+        super(Contact, self).__init__()
+        self._feature_group_synchronize.append(self._call_contacts)
+        self._feature_group_finalize.append(self._finalize_contact)
+
+    def detect_contact_between(self, first_system, second_system):
+        """
+        This method adds contact detection between two objects using the selected contact class.
+        You need to input the two objects that are to be connected.
+
+        Parameters
+        ----------
+        first_system : object
+            Rod or rigid body object
+        second_system : object
+            Rod, rigid body or surface object
+
+        Returns
+        -------
+
+        """
+        sys_idx = [None] * 2
+        for i_sys, sys in enumerate((first_system, second_system)):
+            sys_idx[i_sys] = self._get_sys_idx_if_valid(sys)
+
+        # Create _Contact object, cache it and return to user
+        _contact = _Contact(*sys_idx)
+        self._contacts.append(_contact)
+
+        return _contact
+
+    def _finalize_contact(self):
+
+        # dev : the first indices stores the
+        # (first_rod_idx, second_rod_idx)
+        # to apply the contacts to
+        # Technically we can use another array but it its one more book-keeping
+        # step. Being lazy, I put them both in the same array
+        self._contacts[:] = [(*contact.id(), contact()) for contact in self._contacts]
+
+        # check contact order
+        for (
+            first_sys_idx,
+            second_sys_idx,
+            contact,
+        ) in self._contacts:
+            contact._check_order(
+                self._systems[first_sys_idx],
+                self._systems[second_sys_idx],
+            )
+
+    def _call_contacts(self, time, *args, **kwargs):
```

Comment:
lets keep a practice of adding type hints wherever possible. having them is always useful and doesn't cause any harm

Reply:
this does not time actually, so let me remove it